app/controllers/api/v1/entries_controller.rb

```diff
@@ -73,6 +74,27 @@ def upload_tsv
 
   private
 
+  def authenticate_token
+    auth_header = request.headers['Authorization']
+
+    if auth_header.present? && auth_header.starts_with?('Bearer ')
+      token = AccessToken.find_by(token: auth_header.split(' ').last)
+
+      if token&.live?
+        sign_in(token.user)
+      else
+        render_token_invalid_error
+      end
+    else
+      render_token_invalid_error
+    end
+  end
```

Comment:
tokenメソッドがあるといいかもしれません。


どうやらパターンマッチングを使うと次のようにベアラートークンがとれるようです。

```ruby
def token_in(headers)
  case headers['Authorization']
  in /^Bearer (.+)$/
    Regexp.last_match(1)
   else
     nil
   end
end
```
これを使って次のようなtokenメソッドを定義すると

```ruby
def token
    bearer_token = bearer_token_in request.headers
    AccessToken.find_by(token: bearer_token) if bearer_token
end
```

```suggestion
  def authenticate_token
      if token&.live?
        sign_in(token.user)
      else
        render_token_invalid_error
      end
  end
```

のように書けそうです。

Reply:
こちらで修正しました。
https://github.com/pubannotation/pubdictionaries/pull/142/commits/3b10ed3618da573b4ca182868cf82e8f0eb06f73

また、ロジックを移動できそうだったのでAccessTokenモデルに移動してみました。
https://github.com/pubannotation/pubdictionaries/pull/142/commits/f0505ecfd245e1e33ae1b3f66fc5b88ddb72720c

Reply:
HTTPヘッダーからトークンを取り出す処理は、モデルまで持っていきたくないです。
HTTPリクエストの処理はコントローラーの責務だと思います。

Reply:
確かにリクエスト関係の処理なのでコントローラーの責務ですね..!
修正を元に戻しました。

https://github.com/pubannotation/pubdictionaries/pull/142/commits/00b3a29c10f29d31df8b932d66e29ac9793ef076